### PR TITLE
Editor: fix publish without email validation

### PR DESCRIPTION
**Pull request by gyoshev**
I decided to take a shot at fixing #2650, as it seems to be bothering a lot of people.

It's a bit scary to make such a big change, so I decided to ask beforehand. Judging from #2650 and #4744, it seems best for the email verification notice to be removed, and request email verification only when necessary. be3894b9 removes it, and 501c052 adds a dialog that prompts users to verify their email when they try to publish. As @folletto said, this doesn't break the user flow, and users are more compelled to verify their email when their post is finished.

A few questions:
1. Is a single dialog enough? Are there more places where an email verification is required? It may be too early to remove the notice box.
2. Architecturally, is this the proper place for the email verification dialog, or should it be moved on a more general level?
3. Is it better to reuse translated messages from prior components, rather than introducing new ones? It may not always be feasible, especially when introducing new UI, but is that something that one should strive for?

The dialog still requires some love (more feedback while the dialog actions execute), but I wanted to ask if I'm on the right track.

### CODE REVIEW

**Comment by folletto** [comment]
> be3894b removes it, and 501c052 adds a dialog that prompts users to verify their email when they try to publish.

Question: the dialog should be removed from the Editor, but I'm not sure it's ideal to remove it anywhere else — at least until we have a proper substitute in place. 

Is this PR removing it from just the editor, or everywhere?

> 1. Is a single dialog enough? Are there more places where an email verification is required? It may be too early to remove the notice box.

I think this is related to my question above. So yes: the dialog is the only thing that should be on the Editor, but the notice at the top should stay on all the other screens.

> 2. Architecturally, is this the proper place for the email verification dialog, or should it be moved on a more general level?

cc @mtias 

> 3. Is it better to reuse translated messages from prior components, rather than introducing new ones? It may not always be feasible, especially when introducing new UI, but is that something that one should strive for?

Afaik we have ongoing translations, so the rationale I'd use is: if the change is just a really really minor change (like, a comma with no semantic value) then maybe it could wait. Otherwise I'd never avoid a change because translations will catch up. Maybe let's just be considerate on not changing the same string every day. ;)

**Comment by gyoshev** [comment]
@folletto thanks for the feedback! I understand the considerations around translation, sounds reasonable.

I amended the removal of the email verification notice. Now it is only hidden in the post-editor.

How do you envision the workflow when the user clicks the "I clicked on the email, try again"? Is it okay to just close the window and require another click on "Publish", or something more automat(t)ic?

**Comment by folletto** [comment]
> How do you envision the workflow when the user clicks the "I clicked on the email, try again"? Is it okay to just close the window and require another click on "Publish", or something more automat(t)ic?

Excellent point.

I think that publishing right after clicking the blue confirmation button on the popup is ideal, but the text needs to be clear about it. However I also want to avoid to have a label that gets too long.

Ideas:

1. I clicked on the email, publish now
2. I verified my email, publish
3. Email verified! Try again publishing
4. Email verified! Publish now

Or something around these lines. Of the above ones, I'd probably go for (2).

In that case however I'm wondering if we need to add a button to just dismiss the overlay, or clicking on the background would be enough.

**Comment by hoverduck** [comment]
> In that case however I'm wondering if we need to add a button to just dismiss the overlay, or clicking on the background would be enough.

I'd be fine with either option, but there needs to be some way to dismiss the overlay.  I felt trapped once I opened it and couldn't get rid of it.  I also expected it to disappear when I clicked "Send Email Again".  If I had to choose, I'd say put an X in the top right so it's explicit.

In general I think it looks great, and definitely addresses the original problem.  Just a couple more suggestions:
* I would rephrase the line "click the link below" to say "click the button" instead, since that's what it is.
* I'm not sure where to fit it in, but it would be nice to point out that the post has been saved as a draft so the user doesn't have anxiety that their work will be lost if they can't find the e-mail.

**Comment by folletto** [comment]
> If I had to choose, I'd say put an X in the top right so it's explicit.

The suggestion is good, but our design patterns don't use that. The dialog just have buttons at the bottom, so instead of an `×` we should do a "Cancel" button... which however would mean three buttons. 

So, this could be an alternative (incorporating also the other feedbacks):

screen shot 2016-04-25 at 15 48 14

How is it?

**Comment by shaunandrews** [comment]
I don't think its a good idea to hide the fact that whatever the user writes, they can't publish it.

I think a better alternative is to 1) disable the publish button; and 2) add a persistent (but unobtrusive) notice.

screen shot 2016-04-25 at 11 26 49 am

The notice could be styled to feel more at home with the rest of the editor's sidebar controls, perhaps even displaying more information and actions when clicked:

screen shot 2016-04-25 at 11 27 02 am

--

I've been thinking about the email confirmation (nee verification or activation) process a lot over the past week or so. I think the best solution is to actually remove the limitation on publishing, in favor of making the site private by default. This would let users publish to their site without an interference, but no one else would be able to view the site until they confirm their email address.

--

@meremagee brought up another point:

> paid sites get green flagged and don’t need to verify before publishing [...] we’ll have to test that we aren’t throwing the isverified check for paid users

**Comment by drw158** [comment]
Instead of an accordion, it could be an inline notice, maybe blue or yellow. I think a notice would be more appropriate.

**Comment by folletto** [comment]
I was ok to not show it – the user likely comes from a screen before that has the message already, so I kinda expect that – but I'm also ok in having it inline.

I'd avoid the accordion too, yes.

I'm not sure on the state of the PR. If it's almost done, we could push that in (since I think it's already a big improvement), and then work on that design or even taking it away as you mention in favour of the private flag (which I guess wouldn't require any additional UI here then)?

**Comment by shaunandrews** [comment]
> I was ok to not show it – the user likely comes from a screen before that has the message already, so I kinda expect that – but I'm also ok in having it inline.

I think its important to tell the user upfront and somewhat in their face that they can't click the bright blue button.

> I'm not sure on the state of the PR. If it's almost done, we could push that in (since I think it's already a big improvement)

I'm not onboard with merging the dialog as-is — I think the notice is a better option all around.

**Comment by gyoshev** [comment]
The "I clicked on the email, try again" button in this PR still needs some work. It needs to check if the user has actually verified their e-mail (fetch the new data from the server), and publish afterwards. It also needs the adjustments noted in the latest dialog mockup.

There seem to be some conflicting opinions about the proper workflow here:

- showing a dialog when the user hits publish (this PR)
- disabling the publish button and showing an accordion underneath
- make the site private by default and allow publishing

I like the first option, as it inserts a required step in the workflow, at the precise time when it is needed.

On the other hand, the notice might distract the user away from completing the post (not unlike the notice above the post that has been hidden).

The third seems to be a major change to me -- does confirming the e-mail make the site public, always? What if the user intends to create a private site (and indicated so via a setting), should there be a distinction between an "explicitly private" and "private due to verification" site?

**Comment by gyoshev** [comment]
I've adjusted the dialog according to the latest design & copy by @folletto, and [added a check](https://github.com/Automattic/wp-calypso/pull/4806/commits/4a37b569404f4368ad5a78e91d8aba8b8a4fa207#diff-914bdeefa512de6f78c9b7988546d6e3R769) if the current site is a VIP site. I believe this takes care of the case that @meremagee has mentioned? Feedback is much appreciated!

**Comment by coreh** [comment]
Hey there @gyoshev, I've opened a new PR (#5475) based on your branch but on this repo (instead of your fork), so that I could build upon it and add a couple more things. To keep things organized, I'm going to close this PR, but let's continue the discussion thread there. 👉 #5475
